File: java/mlt-core/src/main/java/org/maplibre/mlt/metadata/stream/RleEncodedStreamMetadata.java
package org.maplibre.mlt.metadata.stream;

import java.io.IOException;
import me.lemire.integercompression.IntWrapper;
import org.apache.commons.lang3.ArrayUtils;
import org.maplibre.mlt.converter.encodings.EncodingUtils;
import org.maplibre.mlt.decoder.DecodingUtils;

public class RleEncodedStreamMetadata extends StreamMetadata {
  int runs;
  int numRleValues;

  // TODO: refactor -> use builder pattern

  /**
   * Only used for RLE encoded integer values. Not needed for rle encoded boolean and byte values.
   *
   * @param numValues After LogicalLevelTechnique was applied -> numRuns + numValues
   * @param runs Length of the runs array
   * @param numRleValues Used for pre-allocating the arrays on the client for faster decoding
   */
  public RleEncodedStreamMetadata(
      PhysicalStreamType physicalStreamType,
      LogicalStreamType logicalStreamType,
      LogicalLevelTechnique logicalLevelTechnique1,
      LogicalLevelTechnique logicalLevelTechnique2,
      PhysicalLevelTechnique physicalLevelTechnique,
      int numValues,
      int byteLength,
      int runs,
      int numRleValues) {
    super(
        physicalStreamType,
        logicalStreamType,
        logicalLevelTechnique1,
        logicalLevelTechnique2,
        physicalLevelTechnique,
        numValues,
        byteLength);
    this.runs = runs;
    this.numRleValues = numRleValues;
  }

  public byte[] encode() throws IOException {
    final var encodedRleInfo =
        EncodingUtils.encodeVarints(new int[] {runs, numRleValues}, false, false);
    return ArrayUtils.addAll(super.encode(), encodedRleInfo);
  }

  public static RleEncodedStreamMetadata decode(byte[] tile, IntWrapper offset) throws IOException {
    var streamMetadata = StreamMetadata.decode(tile, offset);
    var rleInfo = DecodingUtils.decodeVarints(tile, offset, 2);
    return new RleEncodedStreamMetadata(
        streamMetadata.physicalStreamType(),
        streamMetadata.logicalStreamType(),
        streamMetadata.logicalLevelTechnique1(),
        streamMetadata.logicalLevelTechnique2(),
        streamMetadata.physicalLevelTechnique(),
        streamMetadata.numValues(),
        streamMetadata.byteLength(),
        rleInfo[0],
        rleInfo[1]);
  }

  public static RleEncodedStreamMetadata decodePartial(
      StreamMetadata streamMetadata, byte[] tile, IntWrapper offset) throws IOException {
    var rleInfo = DecodingUtils.decodeVarints(tile, offset, 2);
    return new RleEncodedStreamMetadata(
        streamMetadata.physicalStreamType(),
        streamMetadata.logicalStreamType(),
        streamMetadata.logicalLevelTechnique1(),
        streamMetadata.logicalLevelTechnique2(),
        streamMetadata.physicalLevelTechnique(),
        streamMetadata.numValues(),
        streamMetadata.byteLength(),
        rleInfo[0],
        rleInfo[1]);
  }

  public int runs() {
    return this.runs;
  }

  public int numRleValues() {
    return this.numRleValues;
  }
}

File: ts/src/metadata/tile/rleEncodedStreamMetadata.ts
import { decodeStreamMetadata, type StreamMetadata } from "./streamMetadata";
import { decodeVarintInt32 } from "../../decoding/integerDecodingUtils";
import type IntWrapper from "../../decoding/intWrapper";

export function decodeRleEncodedStreamMetadata(tile: Uint8Array, offset: IntWrapper): RleEncodedStreamMetadata {
    const streamMetadata = decodeStreamMetadata(tile, offset);
    return decodePartialRleEncodedStreamMetadata(streamMetadata, tile, offset);
}

export function decodePartialRleEncodedStreamMetadata(
    streamMetadata: StreamMetadata,
    tile: Uint8Array,
    offset: IntWrapper,
): RleEncodedStreamMetadata {
    const rleInfo = decodeVarintInt32(tile, offset, 2);
    return {
        physicalStreamType: streamMetadata.physicalStreamType,
        logicalStreamType: streamMetadata.logicalStreamType,
        logicalLevelTechnique1: streamMetadata.logicalLevelTechnique1,
        logicalLevelTechnique2: streamMetadata.logicalLevelTechnique2,
        physicalLevelTechnique: streamMetadata.physicalLevelTechnique,
        numValues: streamMetadata.numValues,
        byteLength: streamMetadata.byteLength,
        decompressedCount: rleInfo[1],
        runs: rleInfo[0],
        numRleValues: rleInfo[1],
    };
}

export type RleEncodedStreamMetadata = StreamMetadata & {
    readonly runs: number;
    readonly numRleValues: number;
};
